matteodelabre/toltec
# Add uninstall script and make the bootstrap script install-only

This PR adds a new `toltecctl` script, installed by the `toltec-bootstrap` package, that supersedes the former `entware-reenable` script and provides a way to uninstall Toltec (fixes #295). This script could be extended to add new install management features in the future. The uninstall script does the following things:

* Uninstalls all Toltec and Entware packages, running removal scripts for each package.
* Removes the bind mount of Toltec files on /opt (the mount is removed lazily so that it does not fail even if some processes that have files opened in that folder are still running)
* Removes all Toltec files
* Clears Toltec definitions from .bashrc
* Re-enables xochitl
* Reboots the device

Another major change of this PR is to make the bootstrap script fail if files from previous installs remain in the folders used by Toltec (fixes #332). With this change, the bootstrap script only has a single role, that is to create a new Toltec install in a clean system. Roles previously assumed explicitly or implicitly by the bootstrap script are deprecated:

* To re-enable an existing install after a system update, users should run `toltecctl enable` instead of re-running the bootstrap script.
* To create a new clean install, users should first uninstall the existing one and then run the bootstrap script.
* To switch to another branch (see #341), users should manually edit the `/opt/etc/opkg.conf` file instead of re-running the bootstrap script (a future PR could extend the `toltecctl` script to handle switching branches).

Test plan
=========

(Only tested on rM2 so far.)

Uninstall script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Run `~/.local/bin/toltecctl enable` and re-source .bashrc (to add .local/bin to the PATH)
* Run `toltecctl uninstall`
* After the tablet reboots, make sure no traces of the Toltec install remain (in ~/.entware, /opt, /lib/systemd/system/opt.mount)
* If you had disabled xochitl, make sure that it starts up again after uninstalling

Re-enable script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Simulate a system update, for example by switching to the other root partition or using [remarkable-update](https://github.com/ddvk/remarkable-update)
* Run `~/.local/bin/toltecctl enable`
* Make sure that Toltec has been re-enabled (files visible in /opt, opkg command working, apps running)

Bootstrap script
----------------

* Copy the new bootstrap script to the tablet.
* Try running the bootstrap script with an existing Toltec install (the script should fail and tell you that Toltec is already installed)
* Try running the bootstrap script with only some parts of a Toltec install remaining (for example, only the .entware folder - the script should fail and tell you which files need to be cleaned up)
* Run the bootstrap script after uninstalling Toltec or from a clean system, and make sure the install works.
